grid/lib/shared_variable.py

```diff
@@ -0,0 +1,339 @@
+        W3_alice, W3_bob = share(W3)
+        W5_alice, W5_bob = share(W5)
+
+        swap_shares(W0_bob, party, interface)
+        swap_shares(W1_bob, party, interface)
+        swap_shares(W3_bob, party, interface)
+        swap_shares(W5_bob, party, interface)
+
+        quad_alice = [W0_alice, W1_alice, W3_alice, W5_alice]
+        return quad_alice
+    elif (party == 1):
+        W0_bob = swap_shares(torch.LongTensor(
+            x.shape).zero_(), party, interface)
+        W1_bob = swap_shares(torch.LongTensor(
+            x.shape).zero_(), party, interface)
+        W3_bob = swap_shares(torch.LongTensor(
+            x.shape).zero_(), party, interface)
+        W5_bob = swap_shares(torch.LongTensor(
+            x.shape).zero_(), party, interface)
+        quad_bob = [W0_bob, W1_bob, W3_bob, W5_bob]
+        return quad_bob
+
+
+def spdz_sigmoid(x, party, interface):
+    W0, W1, W3, W5 = generate_sigmoid_shares_communication(x, party, interface)
+    x2 = spdz_mul(x, x, party, interface)
+    x3 = spdz_mul(x, x2, party, interface)
+    x5 = spdz_mul(x3, x2, party, interface)
+    temp5 = spdz_mul(x5, W5, party, interface)
+    temp3 = spdz_mul(x3, W3, party, interface)
+    temp1 = spdz_mul(x, W1, party, interface)
+    temp53 = spdz_add(temp5, temp3)
+    temp531 = spdz_add(temp53, temp1)
+    return spdz_add(W0, temp531)
+
+
+class SharedAdd(Function):
+
+    @staticmethod
+    def forward(ctx, a, b):
+        return spdz_add(a, b)
+
+    @staticmethod
+    def backward(ctx, grad_out):
+        grad_out = grad_out.data
+        return grad_out, grad_out
+
+
+class SharedMult(Function):
+
+    @staticmethod
+    def forward(ctx, a, b, party, interface):
+        ctx.save_for_backward(a, b)
+        return spdz_mul(a, b, party, interface)
+
+    @staticmethod
+    def backward(ctx, grad_out, party, interface):
+        a, b = ctx.saved_tensors
+        grad_out = grad_out
+        return Variable(spdz_mul(grad_out.data, b, party, interface)), Variable(spdz_mul(grad_out.data, a, party, interface))
+
+
+class SharedMatmul(Function):
+
+    @staticmethod
+    def forward(ctx, a, b, party, interface):
+        ctx.save_for_backward(a, b)
+        return spdz_matmul(a, b, party, interface)
+
+    @staticmethod
+    def backward(ctx, grad_out, party, interface):
+        a, b = ctx.saved_tensors
+        return spdz_matmul(grad_out,  b.t_(), party, interface), spdz_matmul(grad_out, a.t_(), party, interface)
+
+
+class SharedSigmoid(Function):
+
+    @staticmethod
+    def forward(ctx, a, party, interface):
+        ctx.save_for_backwards(a)
+        return spdz_sigmoid(a, party, interface)
+
+    @staticmethod
+    def backward(ctx, grad_out, party, interface):
+        a = ctx.saved_tensors
+        ones = encode(torch.FloatTensor(a.shape).one_())
+        return spdz_mul(a, public_add(ones, -a, party), party, interface)
+
+
+class SharedVariable(object):
+
+    def __init__(self, var, party, interface, requires_grad=True):
```

Comment:
Can you add a `volatile` argument here?

Reply:
Rather than adding volatile argument made it only accept a variable. This will work well with switch to pytorch 4